### PR TITLE
Theming

### PR DESCRIPTION
**Pull request by aaronshekey**
Let's unify our network theming within Stacks, shall we? This PR explores generating a palette from a single primary color's H/S/L values and provides a few atomic classes, and a ton of CSS variables.

- [x] Create basic CSS-native color algorithms
- [x] Document CSS variables
- [x] Theme each component
- [x] Decide if we just hide _how_ the colors are made and just unify on `theme-primary-[xx]` through `theme-primary-[xxx]`.
- [x] Also include `theme-secondary` for Enterprise and the like
- [x] Refine the color algorithms
- [x] Improve dark mode a bit
- [x] Figure out n themes per page
- [x] Enable `theme-light__forced` and `theme-dark__forced`
- [x] Fixes for `theme-system`
- [x] Explore dark mode hue overrides
- [x] Better color docs examples
- [x] Document the separate dark mode colors
- [x] Finesse the colors to closer match production

image

### CODE REVIEW

File: docs/assets/less/stacks-documentation.less [hunk 1]
```diff
@@ -388,6 +406,53 @@
...
+    --theme-primary-color-r: 77;
+    --theme-primary-color-g: 168;
+    --theme-primary-color-b: 156;
+
+    --theme-secondary-color-h: 259;
+    --theme-secondary-color-s: 29%;
+    --theme-secondary-color-l: 55%;
+    --theme-secondary-color-r: 129;
+    --theme-secondary-color-g: 108;
+    --theme-secondary-color-b: 174;
```
Comment (g3rv4) [inline docs/assets/less/stacks-documentation.less]: what is this secondary color used for? on Enterprise we have primary, link color, and topbar background color. How is this used?

**Review by g3rv4 — COMMENTED** [review]
Things look good... but it sounds like if the secondary color is a new thing, we may need to update the theme editor on Enterprise to set it.

Is there any reason why this can't wait until we fork 2021.1? I'm afraid we'll need to work some work there and won't have lots of time to test it and adjust things.

File: docs/assets/less/stacks-documentation.less (hunk 1 above)
Comment (aaronshekey) [reply docs/assets/less/stacks-documentation.less]: It's mildly tricky, but consider Stack Overflow. Our primary color is that orange, but our secondary color that gets applied to primary buttons is our blue. This is quite common across the network themes. In a theme like Academia, I'd likely apply the primary color to all the buttons, but the secondary to link colors. This abstracts things just a little bit.

**Comment by aaronshekey** [comment]
Happy to wait on this, @g3rv4.

File: docs/assets/less/stacks-documentation.less (hunk 1 above)
Comment (g3rv4) [reply docs/assets/less/stacks-documentation.less]: oh, so there won't be a way to say "I want links to have this color". That's alright, I just think a PM should ensure that's good for Enterprise.

Comment (paintedbicycle) [reply docs/assets/less/stacks-documentation.less]: I wonder if on Enterprise, instead of unique colours for those UI elements, we only allow the person making the theme to set each of those to either the primary or secondary colour, since that's usually how we do it anyway. Plus black sometimes too

Comment (g3rv4) [reply docs/assets/less/stacks-documentation.less]: we'd need to validate that Enterprise customers are not using the theme editor though... we know that we give them a json file, we don't know what they use after that (and we have no visibility into that)

**Comment by g3rv4** [comment]
happy that you're not waiting for me anymore :P

**Comment by aaronshekey** [comment]
That guy kinda fell off, didn't he.

**Comment by aaronshekey** [comment]
This is currently blocked on solving multiple themes per page. When someone has multiple teams and they land on `stackoverflow.com`, they're greeted by previews of their Teams activity along with public SO. Any user with a Team will be shown at least one page with multiple themes on it.

If we expect to be able to use `fc-primary-400` and for it to have different colors in different context, we'll need a way to scope things. It's not as simple as:

```
body {
    --theme-primary-color-h: 172;
    --theme-primary-color-s: 37%;
    --theme-primary-color-l: 48%;
    --theme-primary-color-r: 77;
    --theme-primary-color-g: 168;
    --theme-primary-color-b: 156;

    --theme-secondary-color-h: 259;
    --theme-secondary-color-s: 29%;
    --theme-secondary-color-l: 55%;
    --theme-secondary-color-r: 129;
    --theme-secondary-color-g: 108;
    --theme-secondary-color-b: 174;

    .theme-[namespace] {
        --theme-primary-color-h: 172;
        --theme-primary-color-s: 37%;
        --theme-primary-color-l: 48%;
        --theme-primary-color-r: 77;
        --theme-primary-color-g: 168;
        --theme-primary-color-b: 156;

        --theme-secondary-color-h: 259;
        --theme-secondary-color-s: 29%;
        --theme-secondary-color-l: 55%;
        --theme-secondary-color-r: 129;
        --theme-secondary-color-g: 108;
        --theme-secondary-color-b: 174;
    }
}
```

**Comment by aaronshekey** [comment]
Did some digging. In order for this to work, we can't just change the individual h, s, and l values per scope, we have to redefine _every_ assembled color variable in each scope. So, on `n` teams, we'd also need to include the following:

```
.team-theme {
    --theme-primary-color: hsl(var(--theme-primary-color-h), var(--theme-primary-color-s), var(--theme-primary-color-l));
    --theme-primary-900: hsl(var(--theme-primary-color-h), var(--theme-primary-color-s), calc(var(--theme-primary-color-l) - 26%));
    --theme-primary-800: hsl(var(--theme-primary-color-h), var(--theme-primary-color-s), calc(var(--theme-primary-color-l) - 21%));
    --theme-primary-700: hsl(var(--theme-primary-color-h), var(--theme-primary-color-s), calc(var(--theme-primary-color-l) - 16%));
    --theme-primary-600: hsl(var(--theme-primary-color-h), var(--theme-primary-color-s), calc(var(--theme-primary-color-l) - 10%));
    --theme-primary-500: hsl(var(--theme-primary-color-h), var(--theme-primary-color-s), calc(var(--theme-primary-color-l) - 5%));
    --theme-primary-400: var(--theme-primary-color);
    --theme-primary-350: rgb(calc(25.5 + var(--theme-primary-color-r) * 0.9), calc(25.5 + var(--theme-primary-color-g) * 0.9), calc(25.5 + var(--theme-primary-color-b) * 0.9));
    --theme-primary-300: rgb(calc(89.25 + var(--theme-primary-color-r) * 0.65), calc(89.25 + var(--theme-primary-color-g) * 0.65), calc(89.25 + var(--theme-primary-color-b) * 0.65));
    --theme-primary-200: rgb(calc(153 + var(--theme-primary-color-r) * 0.4), calc(153 + var(--theme-primary-color-g) * 0.4), calc(153 + var(--theme-primary-color-b) * 0.4));
    --theme-primary-150: rgb(calc(178.5 + var(--theme-primary-color-r) * 0.3), calc(178.5 + var(--theme-primary-color-g) * 0.3), calc(178.5 + var(--theme-primary-color-b) * 0.3));
    --theme-primary-100: rgb(calc(198.9 + var(--theme-primary-color-r) * 0.22), calc(198.9 + var(--theme-primary-color-g) * 0.22), calc(198.9 + var(--theme-primary-color-b) * 0.22));
    --theme-primary-075: rgb(calc(216.75 + var(--theme-primary-color-r) * 0.15), calc(216.75 + var(--theme-primary-color-g) * 0.15), calc(216.75 + var(--theme-primary-color-b) * 0.15));
    --theme-primary-050: rgb(calc(234.6 + var(--theme-primary-color-r) * 0.08), calc(234.6 + var(--theme-primary-color-g) * 0.08), calc(234.6 + var(--theme-primary-color-b) * 0.08));
    --theme-primary-025: rgb(calc(244.8 + var(--theme-primary-color-r) * 0.04), calc(244.8 + var(--theme-primary-color-g) * 0.04), calc(244.8 + var(--theme-primary-color-b) * 0.04));
    --theme-primary-color-fade-50: hsla(var(--theme-primary-color-h), var(--theme-primary-color-s), var(--theme-primary-color-l), 0.5);
    --theme-primary-color-fade-75: hsla(var(--theme-primary-color-h), var(--theme-primary-color-s), var(--theme-primary-color-l), 0.75);
    --theme-primary-color-fade-80: hsla(var(--theme-primary-color-h), var(--theme-primary-color-s), var(--theme-primary-color-l), 0.8);
    --theme-primary-color-fade-15: hsla(var(--theme-primary-color-h), var(--theme-primary-color-s), var(--theme-primary-color-l), 0.15);
    --theme-primary-color-fade-35: hsla(var(--theme-primary-color-h), var(--theme-primary-color-s), var(--theme-primary-color-l), 0.35);
    --theme-secondary-color: hsl(var(--theme-secondary-color-h), var(--theme-secondary-color-s), var(--theme-secondary-color-l));
    --theme-secondary-900: hsl(var(--theme-secondary-color-h), var(--theme-secondary-color-s), calc(var(--theme-secondary-color-l) - 26%));
    --theme-secondary-800: hsl(var(--theme-secondary-color-h), var(--theme-secondary-color-s), calc(var(--theme-secondary-color-l) - 21%));
    --theme-secondary-700: hsl(var(--theme-secondary-color-h), var(--theme-secondary-color-s), calc(var(--theme-secondary-color-l) - 16%));
    --theme-secondary-600: hsl(var(--theme-secondary-color-h), var(--theme-secondary-color-s), calc(var(--theme-secondary-color-l) - 10%));
    --theme-secondary-500: hsl(var(--theme-secondary-color-h), var(--theme-secondary-color-s), calc(var(--theme-secondary-color-l) - 5%));
    --theme-secondary-400: var(--theme-secondary-color);
    --theme-secondary-350: rgb(calc(25.5 + var(--theme-secondary-color-r) * 0.9), calc(25.5 + var(--theme-secondary-color-g) * 0.9), calc(25.5 + var(--theme-secondary-color-b) * 0.9));
    --theme-secondary-300: rgb(calc(89.25 + var(--theme-secondary-color-r) * 0.65), calc(89.25 + var(--theme-secondary-color-g) * 0.65), calc(89.25 + var(--theme-secondary-color-b) * 0.65));
    --theme-secondary-200: rgb(calc(153 + var(--theme-secondary-color-r) * 0.4), calc(153 + var(--theme-secondary-color-g) * 0.4), calc(153 + var(--theme-secondary-color-b) * 0.4));
    --theme-secondary-150: rgb(calc(178.5 + var(--theme-secondary-color-r) * 0.3), calc(178.5 + var(--theme-secondary-color-g) * 0.3), calc(178.5 + var(--theme-secondary-color-b) * 0.3));
    --theme-secondary-100: rgb(calc(198.9 + var(--theme-secondary-color-r) * 0.22), calc(198.9 + var(--theme-secondary-color-g) * 0.22), calc(198.9 + var(--theme-secondary-color-b) * 0.22));
    --theme-secondary-075: rgb(calc(216.75 + var(--theme-secondary-color-r) * 0.15), calc(216.75 + var(--theme-secondary-color-g) * 0.15), calc(216.75 + var(--theme-secondary-color-b) * 0.15));
    --theme-secondary-050: rgb(calc(234.6 + var(--theme-secondary-color-r) * 0.08), calc(234.6 + var(--theme-secondary-color-g) * 0.08), calc(234.6 + var(--theme-secondary-color-b) * 0.08));
    --theme-secondary-025: rgb(calc(244.8 + var(--theme-secondary-color-r) * 0.04), calc(244.8 + var(--theme-secondary-color-g) * 0.04), calc(244.8 + var(--theme-secondary-color-b) * 0.04));
    --theme-secondary-color-fade-50: hsla(var(--theme-secondary-color-h), var(--theme-secondary-color-s), var(--theme-secondary-color-l), 0.5);
    --theme-secondary-color-fade-75: hsla(var(--theme-secondary-color-h), var(--theme-secondary-color-s), var(--theme-secondary-color-l), 0.75);
    --theme-secondary-color-fade-80: hsla(var(--theme-secondary-color-h), var(--theme-secondary-color-s), var(--theme-secondary-color-l), 0.8);
    --theme-secondary-color-fade-15: hsla(var(--theme-secondary-color-h), var(--theme-secondary-color-s), var(--theme-secondary-color-l), 0.15);
    --theme-secondary-color-fade-35: hsla(var(--theme-secondary-color-h), var(--theme-secondary-color-s), var(--theme-secondary-color-l), 0.35);

    --theme-primary-color-h: 172;
    --theme-primary-color-s: 37%;
    --theme-primary-color-l: 48%;
    --theme-primary-color-r: 77;
    --theme-primary-color-g: 168;
    --theme-primary-color-b: 156;

    --theme-secondary-color-h: 259;
    --theme-secondary-color-s: 29%;
    --theme-secondary-color-l: 55%;
    --theme-secondary-color-r: 129;
    --theme-secondary-color-g: 108;
    --theme-secondary-color-b: 174;
}
```

Not great!

**Comment by dancormier** [comment]
I have an idea of how to address the multiple theme issue.🤞

tl;dr: [codepen proof of concept](https://codepen.io/dancormier/pen/d0fa7bd79e619fe312f6e2e188cf4e87?editors=0100).

### Hypothesis

Wrapping all rules in both a base specifier and a base+nested specifier creates separate specificities to act upon, which will allow us to redefine base variables that will cascade within their respective scopes.

Essentially, every variable created within `body, body .nested-theme` gets two references -- one in `body` and one in `body .nested-theme`, which can be acted upon individually.

### Example

Note: For simplicity sake, I created some dummy variables and class names that aren't necessarily reflected in Stacks, but the implementation should translate.

```CSS
.themed, .themed .nested-theme {
  --theme-primary-color: hsl(var(--theme-primary-color-h), var(--theme-primary-color-s), var(--theme-primary-color-l));
  --theme-secondary-color: hsl(var(--theme-secondary-color-h), var(--theme-secondary-color-s), var(--theme-secondary-color-l));

  background-color: var(--theme-secondary-color);
  border: 3px solid var(--theme-primary-color);
  color: var(--theme-primary-color);
  padding: 1rem;
}

.themed {
  --theme-primary-color-h: 0;
  --theme-primary-color-s: 100%;
  --theme-primary-color-l: 50%;
  --theme-secondary-color-h: 100;
  --theme-secondary-color-s: 70%;
  --theme-secondary-color-l: 90%;

  .nested-theme {
    --theme-primary-color-h: 200;
    --theme-secondary-color-h: 300;
  }
}
```

```HTML
<div class="themed">
  Base theme
  <div class="nested-theme">Custom theme</div>
</div>
```

The above results in something like this:

**Comment by aaronshekey** [comment]
Now that @dancormier solved the scoping issues, this is now documented and ready to go.

Though, further down the road we could do further theming on the following components to include the secondary colors:

- [x] `s-link`
- [x] `s-btn`
- [x] `s-activity-indicator`
- [x] `s-checkbox`
- [x] `s-radio`
- [x] `s-input` selection / focus states
- [x] `s-progress`
- [x] `s-select` 
- [x] `s-textarea`
- [x] `s-toggle-switch`
- [x] `s-user-card`

**Comment by aaronshekey** [comment]
We still need a consistent way of doing `.theme-light__forced` and `.theme-dark__forced`.

So so so close.

Also, took at look at getting this into production. In order for our first-class Stacks components to be themed, we'll need to provide theming CSS variables for each of our network sites.

@NickCraver Might look for your advice on shipping something like this while we transition to theming in the client: 

```
<style>
    --theme-primary-color-h: 172;
    --theme-primary-color-s: 37%;
    --theme-primary-color-l: 48%;
    --theme-primary-color-r: 77;
    --theme-primary-color-g: 168;
    --theme-primary-color-b: 156;
    --theme-secondary-color-h: 259;
    --theme-secondary-color-s: 29%;
    --theme-secondary-color-l: 55%;
    --theme-secondary-color-r: 129;
    --theme-secondary-color-g: 108;
    --theme-secondary-color-b: 174;
</style>
```

**Comment by aaronshekey** [comment]
We're just down to me choosing slightly different color curves to closer match what we got for oranges and blues. Then I'll update the docs to talk about how we can set different dark mode colors entirely. This looks functionally finished. 🎉

**Comment by aaronshekey** [comment]
Hey, this one took us less than a year!
